### ISSUE TITLE
weird bundle exec - nohup behavior where SIGHUP gets raised to child process

### ISSUE THREAD

**Issue by oneamtu** [bundler/bundler#6150]
At the company I work for we ran into this weird problem when running processes through `nohup` and `bundle exec`. `nohup` is a utility that is supposed to trap SIGHUPs from bubbling to a child process, thus allowing a user to daemonize a command by running it with `nohup` and then closing the parent shell. 

For whatever reason, with more recent versions of bundler past 1.12.5 SIGHUPs bubble up to the child process. 1.12.5 works great. We also found a workaround with more recent versions of bundler where the ordering of the commands changes the behavior - `nohup bundle exec` fails and `bundle exec nohup` works. Here's an example of the test which I uploaded here - https://github.com/oneamtu/bundler_nohup_test:

```
[19:56:02] oneamtu:test $ nohup rake loop_de_loop >> loop.log 2>&1 &
[1] 7553
[19:56:27] oneamtu:test $ kill -hup 7553
[19:56:34] oneamtu:test $ kill -kill 7553
[1]  + 7553 killed     nohup rake loop_de_loop >> loop.log 2>&1
[19:56:44] oneamtu:test $ nohup bundle exec rake loop_de_loop >> loop.log 2>&1 &
[1] 7657
[19:56:52] oneamtu:test $ kill -hup 7657
[1]  + 7657 exit 1     nohup bundle exec rake loop_de_loop >> loop.log 2>&1
[19:56:59] oneamtu:test $ bundle exec nohup rake loop_de_loop >> loop.log 2>&1 &
[1] 7746
[19:57:12] oneamtu:test $ kill -hup 7746
[19:57:17] oneamtu:test $ kill -kill 7746
[1]  + 7746 killed     bundle exec nohup rake loop_de_loop >> loop.log 2>&1

[6:09:53] oneamtu:test $ nohup --version
nohup (GNU coreutils) 8.25
Copyright (C) 2016 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Written by Jim Meyering.
```

## Environment

```
Bundler       1.16.0
  Platforms   ruby, x86_64-linux
Ruby          2.4.2p198 (2017-09-14 revision 59899) [x86_64-linux]
  Full Path   /home/oneamtu/.asdf/installs/ruby/2.4.2/bin/ruby
  Config Dir  /home/oneamtu/.asdf/installs/ruby/2.4.2/etc
RubyGems      2.6.13
  Gem Home    /home/oneamtu/.asdf/installs/ruby/2.4.2/lib/ruby/gems/2.4.0
  Gem Path    /home/oneamtu/.gem/ruby/2.4.0:/home/oneamtu/.asdf/installs/ruby/2.4.2/lib/ruby/gems/2.4.0
  User Path   /home/oneamtu/.gem/ruby/2.4.0
  Bin Dir     /home/oneamtu/.asdf/installs/ruby/2.4.2/bin
Tools
  Git         2.7.4
  RVM         not installed
  rbenv       not installed
  chruby      not installed
```

## Bundler Build Metadata

```
Built At          2017-10-31
Git SHA           10f20fa33
Released Version  true
```

## Bundler settings

```
path
  Set for your local app (/home/oneamtu/test/test/.bundle/config): "vendor/bundle"
disable_shared_gems
  Set for your local app (/home/oneamtu/test/test/.bundle/config): true
```

### Gemfile

```ruby
source 'https://rubygems.org'

gem 'rake', '~> 11.0'
```

### Gemfile.lock

```
GEM
  remote: https://rubygems.org/
  specs:
    rake (11.3.0)
```

**Comment by colby-swandale** [comment]
Thanks for the detailed report! A PR fixing this issue would be very welcome 👍

**Comment by atomical** [comment]
What would be the work around for bundle exec executable_name ?  Where the executable name is only available when doing bundle exec?

**Comment by segiddins** [comment]
I wonder if [this change](https://github.com/bundler/bundler/compare/1-12-stable...v1.13.0#diff-fcaa44c40b054ace4b039a38e3ffe82fR73) in 1.13 is responsible

**Comment by shayonj** [comment]
@segiddins I am unable to follow that link, but, I believe it started with this change introducing the unexpectedness with Signals

https://github.com/bundler/bundler/blob/93e11fc7e3b9d067affe637c0e4c264d5b82bcd1/lib/bundler/cli/exec.rb#L63

from:

https://github.com/bundler/bundler/commit/93e11fc7e3b9d067affe637c0e4c264d5b82bcd1#diff-fcaa44c40b054ace4b039a38e3ffe82fR63

Before I proceed, **IMPORTANT NOTE:** This bug only exists when you vendor in your gems (which requires `Bundler::CLI::Exec` to call the executable). If you were to try the steps from description without vendoring, this bug does not exists.
 
**Explanation:**

Like I mentioned, this only happens when you vendor in.

- The issue starts with bundler swallowing `HUP` signal here

https://github.com/bundler/bundler/blob/060e2f953c8f08dd7bed9711863e6341eebe5fe2/lib/bundler/cli/exec.rb#L74

- Now, because `rake` is vendored in (per the description) (and is not executed by `ruby_executable_hooks`), the same is loaded here

https://github.com/bundler/bundler/blob/060e2f953c8f08dd7bed9711863e6341eebe5fe2/lib/bundler/cli/exec.rb#L75

So, when a `HUP` signal is sent, it gets rescued/swalloed well before `nohup` gets a chance to work with it.

This entire phenomenon, can by tried by this simple script:

**test.rb** (

```ruby
IGNORE_SIGNALS = %w[SEGV BUS ILL FPE VTALRM KILL STOP]
signals = Signal.list.keys - IGNORE_SIGNALS
signals.each do |s|
  trap(s, "DEFAULT")
end

sleep 100
```

Now, try using `nohup` on the file:

```
nohup ruby test.rb 
appending output to nohup.out

...

# send a `kill -hup $PID` on the process
```
You will see the nohup died.

Next, add `HUP` to `_` and try again. You will see that `nohup` no longer dies and its not being trapped/swalled anymore.

I will try to cut a PR soon.

**Comment by shayonj** [comment]
I am curious, why does bundler needs to swallow any Signal at all? The changes that were introduced to fix the issue: https://github.com/bundler/bundler/issues/4568, (which I think is the probable cause for this current issue) seems a little unclear to me, so I am curious :). I am also unable to reproduce the original issue from https://github.com/bundler/bundler/issues/4568

```
bundle _1.12.0_ exec puma
Puma starting in single mode...
* Version 3.4.0 (ruby 2.3.3-p222), codename: Owl Bowl Brawl
* Min threads: 5, max threads: 5
* Environment: development
* Listening on tcp://0.0.0.0:3001
Use Ctrl-C to stop
^C- Gracefully stopping, waiting for requests to finish
=== puma shutdown: 2017-12-13 21:20:46 -0800 ===
- Goodbye!
echo $?                  
0
```

I am trying to determine whether it makes sense to add `HUP` to

https://github.com/bundler/bundler/blob/060e2f953c8f08dd7bed9711863e6341eebe5fe2/lib/bundler/cli/exec.rb#L9

Or just not have signals get swallowed by bundler at all, but leave it to to executables to manage it. For example, puma does a lot of custom stuff to handle signals

**Comment by indirect** [comment]
@shayonj the change was to allow signals to be passed to children, but perhaps it is no longer needed? Can you also test this after `bundle config disable_exec_load true`?

@segiddins any thoughts here? we didn't end up reverting the exec/load stuff, did we?

**Comment by shayonj** [comment]
@indirect ah! I believe that makes sense, thanks!

And, yep, I can no longer replicate the issue after setting `disable_exec_load` to `true`

Looks like `disable_exec_load` is in play here

https://github.com/bundler/bundler/blob/060e2f953c8f08dd7bed9711863e6341eebe5fe2/lib/bundler/cli/exec.rb#L27

**Comment by indirect** [comment]
Hmm. So maybe this means we can stop reserving signals in some versions of Ruby? Or maybe Puma changed the way that it traps signals, and our work-around is no longer needed.

**Comment by segiddins** [comment]
Given that we're just setting the signal handlers all back to their default, I wonder if Ruby's default handler for `SIGHUP` is weird?

**Comment by shayonj** [comment]
hm, perhaps I am not following 😅 

I believe along with trapping signal handlers, we are also _swallowing_ them (i.e we are not re-raising them), which is why `HUP` is never caught by `nohup` (in this specific example). Here is a simple example:

**test.rb**
```ruby
IGNORE_SIGNALS = %w[SEGV BUS ILL FPE VTALRM KILL STOP]
signals = Signal.list.keys - IGNORE_SIGNALS
signals.each do |s|
  trap(s, "DEFAULT")
end

Kernel.exec("ruby ./test1.rb")
# Kernel.load("./test1.rb")
```

**test1.rb**
```ruby
Signal.trap("HUP") do
  puts "--> Received hup from test.rb"
end
sleep 100
```

Now, run this via

```
nohup ruby test.rb
```

And kill via 

```
nohup ruby test.rb
```

You will see that, `nohup` died/return, because we are trapping/swallowing `HUP`. If you, add `HUP` to `IGNORE_SIGNALS`, the behavior is now expected and will also see that the signal is being received by executed file (`test1.rb`), in the logs.

----

In short, I am wondering if we need to trap any signals at all. I still can't replicate the puma specific behavior :D.

**Comment by segiddins** [comment]
But all we’re doing is trapping with the default handler, which in theory should be no different than not calling trap at all

**Comment by shayonj** [comment]
Yes, you are right, so I went digging. Turns outs `Signal.trap` in this case is override the `SIG_IGN` handler that `nohup` has set. This can be replicated (easier to) from a linux machine, where you read a Process' status from `/proc`:

**test.rb**

```ruby
p Signal.trap("SIGHUP", "SIG_IGN")
p Signal.trap("SIGHUP", "SYSTEM_DEFAULT")
sleep 100
```

```ruby
ruby test.rb 2>&2 &

[1] 23156
```

Now, the process status dump:

```
cat /proc/23156/status
Name:	ruby
Umask:	0022
State:	S (sleeping)
Tgid:	23156
Ngid:	0
Pid:	23156
PPid:	52
TracerPid:	0
Uid:	0	0	0	0
Gid:	0	0	0	0
FDSize:	256
Groups:
NStgid:	23156
NSpid:	23156
NSpgid:	23156
NSsid:	52
VmPeak:	   40508 kB
VmSize:	   40508 kB
VmLck:	       0 kB
VmPin:	       0 kB
VmHWM:	    8036 kB
VmRSS:	    8036 kB
RssAnon:	    3596 kB
RssFile:	    4440 kB
RssShmem:	       0 kB
VmData:	    5192 kB
VmStk:	    8188 kB
VmExe:	       4 kB
VmLib:	    6624 kB
VmPTE:	      88 kB
VmPMD:	      12 kB
VmSwap:	       0 kB
HugetlbPages:	       0 kB
Threads:	2
SigQ:	0/7753
SigPnd:	0000000000000000
ShdPnd:	0000000000000000
SigBlk:	0000000000000000
SigIgn:	0000000000000000
SigCgt:	00000001c2007e4e
CapInh:	00000000a80425fb
CapPrm:	00000000a80425fb
CapEff:	00000000a80425fb
CapBnd:	00000000a80425fb
CapAmb:	0000000000000000
Seccomp:	2
Cpus_allowed:	f
Cpus_allowed_list:	0-3
Mems_allowed:	1
Mems_allowed_list:	0
voluntary_ctxt_switches:	1
nonvoluntary_ctxt_switches:	8
```

Notice that `SigIgn` is `0000000000000000`, this number is bitmask, in short there is no `SIG_IGN`/`IGNORE` handler present on this process.

I have opened a bug report with ruby and have a working version locally on ruby trunk, lets see what they say (https://bugs.ruby-lang.org/issues/14196). I will open a Github PR for ruby soon.

**Comment by segiddins** [comment]
Thanks. Looks like we might need to re-consider setting everything to `default`, and manually keep a list of all the signals that _bundler itself_ is trapping and just restore those handlers

**Comment by shayonj** [comment]
yes.

>  and manually keep a list of all the signals that bundler itself is trapping and just restore those handlers

Do we know if there are some specific signals that Bundler wish to trap?

If I am getting this right, anything thats not in `IGNORE_SIGNALS` is being being trapped by bundler. Hence, these are the ones we should be trapping and restoring? But, if we wish to restore, then should we even trap them in the first place ? Or, by extension, what are the benefits of trapping
 any signal at all? 😅 🙈 (don't mean to be PITA, apologies, if so)

**Comment by segiddins** [comment]
> Or, by extension, what are the benefits of trapping any signal at all?

We've historically trapped signals such as `SIGINT` to ensure that `^C` properly kills bundler during downloads

**Comment by colby-swandale** [comment]
@segiddins can this be closed?
